File: mdpp.py
#!/usr/bin/env python
#
# Filter that processes given Markdown file and replaces :
#
# - "#(file)" with the content of the file (if path is relative, this
#   is relative to the Markdown file).
# - "?(command)" with the output of the command (running in the Markdown
#   file directory.
#
# Output is written on command line.

import re
import sys
import subprocess


def execute(match):
    process = subprocess.Popen(match.group(1),
                               stdout=subprocess.PIPE,
                               stderr=subprocess.PIPE)
    output, errput = process.communicate()
    if process.returncode != 0:
        print("Error running command '%s':\n%s" % (match, errput))
        sys.exit(2)
    return output.strip()


def readfile(match):
    with open(match.group(1)) as stream:
        return stream.read().strip()


def main(filename):
    with open(filename) as stream:
        content = stream.read()
    content = re.sub(r'^\?\((.+)\)$', execute, content, flags=re.MULTILINE)
    content = re.sub(r'^#\((.+)\)$', readfile, content, flags=re.MULTILINE)
    print(content)


if __name__ == '__main__':
    if len(sys.argv) != 2:
        print('You must pass Markdow file to process on command line')
        sys.exit(1)
    main(sys.argv[1])

File: mdpp.go
package main

import (
	"fmt"
	"io/ioutil"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strings"
)

var directory = ""

func printError(err error, message string) {
	if err != nil {
		println(message)
		os.Exit(2)
	}
}

func execute(cmd string) string {
	command := exec.Command("sh", "-c", cmd)
	command.Dir = directory
	output, err := command.CombinedOutput()
	result := strings.TrimSpace(string(output))
	printError(err, "Error running command '"+cmd+"': "+result)
	return result
}

func include(file string) string {
	content, err := ioutil.ReadFile(filepath.Join(directory, file))
	printError(err, "Error reading source file '"+file+"'")
	return strings.TrimSpace(string(content))
}

func command(s string) string {
	r := regexp.MustCompile("^(.)\\((.+)\\)$")
	m := r.FindStringSubmatch(s)
	cmd := m[1]
	arg := m[2]
	if cmd == "?" {
		return execute(arg)
	} else {
		return include(arg)
	}
}

func process(file string) {
	source, err := ioutil.ReadFile(file)
	printError(err, "Error reading source file")
	directory, err = filepath.Abs(filepath.Dir(file))
	printError(err, "Error getting directory")
	r := regexp.MustCompile("(?m)^(\\?|@)\\(.+\\)$")
	processed := r.ReplaceAllStringFunc(string(source), command)
	fmt.Println(processed)
}

func main() {
	if len(os.Args) != 2 {
		println("You must pass Markdow file to process on command line")
		os.Exit(1)
	}
	process(os.Args[1])
}
